packages/firestore/scripts/assertion-id-tool.ts

```diff
@@ -0,0 +1,372 @@
+/**
+ * @license
+ * Copyright 2025 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/* eslint-disable @typescript-eslint/no-explicit-any */
+/* eslint-disable no-console */
+
+import * as fs from "fs";
+import {getRandomValues} from 'node:crypto';
+import * as path from "path";
+
+import * as ts from "typescript";
+import yargs from "yargs";
+import { hideBin } from "yargs/helpers";
+
+let isVerbose: boolean = false;
+
+function log(message: any): void {
+  if (isVerbose) {
+    console.log(message);
+  }
+}
+
+// Define the names of the functions we are looking for
+const targetFunctionNames: Set<string> = new Set(["fail", "hardAssert"]);
+
+// Interface to store information about found call sites
+interface CallSiteInfo {
+  fileName: string;
+  functionName: string;
+  line: number;
+  character: number;
+  argumentsText: string[]; // Added to store argument text
+  errorMessage: string | undefined;
+  assertionId: string;
+}
+
+/**
+ * Recursively finds all files with .ts extensions in a directory.
+ * @param dirPath The absolute path to the directory to scan.
+ * @returns An array of absolute paths to the found TypeScript files.
+ */
+function getTsFilesRecursive(dirPath: string): string[] {
+  let tsFiles: string[] = [];
+  try {
+    const entries = fs.readdirSync(dirPath, { withFileTypes: true });
+
+    for (const entry of entries) {
+      const fullPath = path.join(dirPath, entry.name);
+
+      if (entry.isDirectory()) {
+        // Ignore node_modules for performance and relevance
+        if (entry.name === 'node_modules') {
+          // log(`Skipping node_modules directory: ${fullPath}`);
+          continue;
+        }
+        // Recursively scan subdirectories
+        tsFiles = tsFiles.concat(getTsFilesRecursive(fullPath));
+      } else if (entry.isFile() && (entry.name.endsWith(".ts"))) {
+        // Exclude declaration files (.d.ts) as they usually don't contain implementation
+        if (!entry.name.endsWith(".d.ts")) {
+          tsFiles.push(fullPath);
+        }
+      }
+    }
+  } catch (error: any) {
+    console.error(`Error reading directory ${dirPath}: ${error.message}`);
```

Comment:
Consider simply letting the exception bubble up, possibly adding some contextual information. This error message could be easy to miss.

Reply:
Good call. Logged and re-threw.